mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,761 @@
+  // no reuse is encountered.
+  uint64_t temporalReuse = 1;
+  for (auto &accessMatrixOpPair : loopAccessMatrices) {
+    temporalReuse = 1;
+    SmallVector<SmallVector<int64_t, 4>, 4> accessMatrix =
+        accessMatrixOpPair.second;
+    for (int i = permutation.size() - 1; i >= 0; i--) {
+      bool isColumnAllZeros = true;
+      for (SmallVector<int64_t, 4> &row : accessMatrix) {
+        if (row[permutation[i]] != 0) {
+          isColumnAllZeros = false;
+          break;
+        }
+      }
+      if (!isColumnAllZeros)
+        break;
+      temporalReuse *= loopIterationCounts[permutation[i]];
+    }
+    // Increase in temporalReuse decreases the cost.
+    cost -= temporalReuse;
+  }
+  return cost;
+}
+
+/// Removes `dstOp` from its current group and inserts it into the `srcOp`'s
+/// group. Updates `groupId` to reflect the changes.
+static void insertIntoReferenceGroup(
+    Operation *srcOp, Operation* dstOp, DenseMap<Operation*, unsigned> &groupId,
+    SmallVector<llvm::SmallSet<Operation *, 8>, 8> &referenceGroups) {
+  referenceGroups[groupId[srcOp]].insert(
+      referenceGroups[groupId[dstOp]].begin(),
+      referenceGroups[groupId[dstOp]].end());
+  referenceGroups.erase(referenceGroups.begin() + groupId[dstOp]);
+  // Insert operation results in same group-id for both instructions.
+  groupId[dstOp] = groupId[srcOp];
+}
+
+/// Groups ops in `loadAndStoreOps` into `referenceGroups` based on whether or
+/// not they exhibit group-temporal or group-spatial reuse with respect to an
+/// affine.for op present at depth `innermostIndex` in the original loop nest.
+///
+/// Please refer Steve Carr et. al for a detailed description.
+/// https://dl.acm.org/doi/abs/10.1145/195470.195557
+static void buildReferenceGroups(
+    SmallVector<Operation *, 8> &loadAndStoreOps,
+    DenseMap<Operation *, SmallVector<SmallVector<int64_t, 4>, 4>>
+        &loopAccessMatrices,
+    DenseMap<Operation *, SmallVector<int64_t, 4>> &constVector,
+    DenseMap<Operation *, unsigned> &elementsSize, unsigned maxDepth,
+    unsigned innermostIndex,
+    SmallVector<llvm::SmallSet<Operation *, 8>, 8> &referenceGroups) {
+
+  // Two accesses ref1 and ref2 belong to the same reference group with respect
+  // to a loop if :
+  // Criteria 1: There exists a dependence l and
+  //     1.1 l is a loop-independent dependence or
+  //     1.2 l's component for `forOp` is a small constant d (|d|<=2) and all
+  //     other entries are zero.
+  // OR
+  // Criteria 2: ref1 and ref2 refer to the same array and differ by at most d1
+  // in the last subscript dimension, where d1 <= cache line size in terms of
+  // array elements. All other subscripts must be identical.
+  //
+  // We start with all accesses having their own group. Thus, if an access is
+  // not part of any group-reuse, it still has it's own group. Doing this takes
+  // care of self-spatial reuse.
+  constexpr unsigned cache_line_size = 64;
```

Comment:
Use LLVM/MLIR style naming - camel back. Please see style guide on MLIR page. Also, this could be a static member in the pass itself so that it's easy to find and later set. `kCacheLineSize`.

Reply:
Done. Thanks.